### PR TITLE
election: revert the unnecessary setting of the lease to avoid the panic

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

https://github.com/tikv/pd/pull/4044/files#diff-1df57c033011699202ef9b7617f0cee905ed0e86445362af5ec416d2aa403b84R224 introduces an unnecessary setting of the lease, which increases the possibility of panic. This PR reverts it and adds more checks.

image

https://ci2.pingcap.net/blue/organizations/jenkins/pd_test/detail/pd_test/2339/pipeline/51/

### What is changed and how it works?

* Revert the unnecessary setting of the lease to avoid the panic
* Add more `nil` checks.
* Add more unit tests.

### Check List

Tests

- Unit test
- Integration test

Related changes

- The root of all evil: https://github.com/tikv/pd/pull/4044

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/election/lease.go [hunk 1]
```diff
@@ -83,17 +86,17 @@ func (l *lease) Close() error {
 // IsExpired checks if the lease is expired. If it returns true,
 // current leader should step down and try to re-elect again.
 func (l *lease) IsExpired() bool {
-	if l == nil {
+	if l == nil || l.expireTime.Load() == nil {
 		return true
 	}
-	if l.expireTime.Load() == nil {
```
Comment (JmPotato) [inline server/election/lease.go:89]: Involved in https://github.com/tikv/pd/commit/7a6a9daaf37e0cf9e444ccc613a6c70bcccdd05c#diff-ce11c8a270a68ea01352751b710c90d380002a3a037cbd39e588e963e24863e0R81, which is not correct. Fixed by this PR.

File: server/election/lease.go [hunk 2]
```diff
@@ -83,17 +86,17 @@ func (l *lease) Close() error {
...
-		return false
-	}
```
Comment (JmPotato) [inline server/election/lease.go:91]: Involved in https://github.com/tikv/pd/commit/7a6a9daaf37e0cf9e444ccc613a6c70bcccdd05c#diff-ce11c8a270a68ea01352751b710c90d380002a3a037cbd39e588e963e24863e0R81, which is not correct. Fixed by this PR.

File: server/election/leadership.go [hunk 3]
```diff
@@ -221,5 +234,4 @@ func (ls *Leadership) Reset() {
 		ls.keepAliceCancelFunc()
 	}
 	ls.getLease().Close()
-	ls.setLease(nil)
```
Comment (JmPotato) [inline server/election/leadership.go:224]: Involved in https://github.com/tikv/pd/pull/4044/files#diff-1df57c033011699202ef9b7617f0cee905ed0e86445362af5ec416d2aa403b84R224.

File: server/election/leadership_test.go [hunk 4]
```diff
@@ -33,7 +33,7 @@ var _ = Suite(&testLeadershipSuite{})
 
 type testLeadershipSuite struct{}
 
-const defaultTestLeaderLease = 1
+const defaultTestLeaderLeaseTimeout = 1
```
Comment (rleungx) [inline server/election/leadership_test.go:36]: make it shorter?

**Comment by rleungx** [comment]
BTW, how much time will the test cost?

**Comment by JmPotato** [comment]
> BTW, how much time will the test cost?

image

About 18s.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by rleungx — APPROVED** [review]

**Pull request merged**